// nano-jsx/lib/bundles/bundle.full.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VERSION = exports.printVersion = void 0;
// core
const core_js_1 = require("../core.js");
const lazy_js_1 = require("../lazy.js");
const helpers_js_1 = require("../helpers.js");
// useful tools
const component_js_1 = require("../component.js");
const fragment_js_1 = require("../fragment.js");
const store_js_1 = require("../store.js");
const context_js_1 = require("../context.js");
const withStyles_js_1 = require("../withStyles.js");
// built-in components
const helmet_js_1 = require("../components/helmet.js");
const link_js_1 = require("../components/link.js");
const img_js_1 = require("../components/img.js");
const visible_js_1 = require("../components/visible.js");
const Router = __importStar(require("../components/router.js"));
// customElement
const customElementsMode_js_1 = require("../customElementsMode.js");
// tagged templates
const jsx_js_1 = require("../jsx.js");
// ui
const UI = __importStar(require("../ui/index.js"));
exports.default = {
    Component: component_js_1.Component,
    Fragment: fragment_js_1.Fragment,
    Helmet: helmet_js_1.Helmet,
    Img: img_js_1.Img,
    Link: link_js_1.Link,
    Router,
    Store: store_js_1.Store,
    UI,
    Visible: visible_js_1.Visible,
    createContext: context_js_1.createContext,
    h: core_js_1.h,
    hydrate: core_js_1.hydrate,
    hydrateLazy: lazy_js_1.hydrateLazy,
    jsx: jsx_js_1.jsx,
    nodeToString: helpers_js_1.nodeToString,
    removeAllChildNodes: core_js_1.removeAllChildNodes,
    render: core_js_1.render,
    task: helpers_js_1.task,
    tick: core_js_1.tick,
    withStyles: withStyles_js_1.withStyles,
    defineAsCustomElements: customElementsMode_js_1.defineAsCustomElements
};
// version
var helpers_js_2 = require("../helpers.js");
Object.defineProperty(exports, "printVersion", { enumerable: true, get: function () { return helpers_js_2.printVersion; } });
var version_js_1 = require("../version.js");
Object.defineProperty(exports, "VERSION", { enumerable: true, get: function () { return version_js_1.VERSION; } });
