// nano-jsx/lib/bundles/bundle.slim.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VERSION = exports.printVersion = void 0;
// core
const core_js_1 = require("../core.js");
const helpers_js_1 = require("../helpers.js");
// useful tools
const component_js_1 = require("../component.js");
const fragment_js_1 = require("../fragment.js");
const store_js_1 = require("../store.js");
const context_js_1 = require("../context.js");
const withStyles_js_1 = require("../withStyles.js");
const Router = __importStar(require("../components/router.js"));
// tagged templates
const jsx_js_1 = require("../jsx.js");
exports.default = {
    Component: component_js_1.Component,
    Fragment: fragment_js_1.Fragment,
    Router,
    Store: store_js_1.Store,
    createContext: context_js_1.createContext,
    h: core_js_1.h,
    jsx: jsx_js_1.jsx,
    render: core_js_1.render,
    task: helpers_js_1.task,
    tick: core_js_1.tick,
    withStyles: withStyles_js_1.withStyles
};
// version
var helpers_js_2 = require("../helpers.js");
Object.defineProperty(exports, "printVersion", { enumerable: true, get: function () { return helpers_js_2.printVersion; } });
var version_js_1 = require("../version.js");
Object.defineProperty(exports, "VERSION", { enumerable: true, get: function () { return version_js_1.VERSION; } });
