// nano-jsx/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VERSION = exports.printVersion = exports.defineAsCustomElements = exports.withStyles = exports.useContext = exports.createContext = exports.Store = exports.Fragment = exports.renderSSR = exports.task = exports.nodeToString = exports.hydrateLazy = exports.jsx = exports.isSSR = exports.Component = exports.tick = exports.hydrate = exports.render = exports.h = void 0;
// core
var core_js_1 = require("./core.js");
Object.defineProperty(exports, "h", { enumerable: true, get: function () { return core_js_1.h; } });
Object.defineProperty(exports, "render", { enumerable: true, get: function () { return core_js_1.render; } });
Object.defineProperty(exports, "hydrate", { enumerable: true, get: function () { return core_js_1.hydrate; } });
Object.defineProperty(exports, "tick", { enumerable: true, get: function () { return core_js_1.tick; } });
// component
var component_js_1 = require("./component.js");
Object.defineProperty(exports, "Component", { enumerable: true, get: function () { return component_js_1.Component; } });
// built-in Components
__exportStar(require("./components/index.js"), exports);
// export some defaults (Nano)
const core_js_2 = require("./core.js");
Object.defineProperty(exports, "isSSR", { enumerable: true, get: function () { return core_js_2.isSSR; } });
const ssr_js_1 = require("./ssr.js");
exports.default = { h: core_js_2.h, render: core_js_2.render, hydrate: core_js_2.hydrate, renderSSR: ssr_js_1.renderSSR, isSSR: core_js_2.isSSR };
var jsx_js_1 = require("./jsx.js");
Object.defineProperty(exports, "jsx", { enumerable: true, get: function () { return jsx_js_1.jsx; } });
var lazy_js_1 = require("./lazy.js");
Object.defineProperty(exports, "hydrateLazy", { enumerable: true, get: function () { return lazy_js_1.hydrateLazy; } });
var helpers_js_1 = require("./helpers.js");
Object.defineProperty(exports, "nodeToString", { enumerable: true, get: function () { return helpers_js_1.nodeToString; } });
Object.defineProperty(exports, "task", { enumerable: true, get: function () { return helpers_js_1.task; } });
var ssr_js_2 = require("./ssr.js");
Object.defineProperty(exports, "renderSSR", { enumerable: true, get: function () { return ssr_js_2.renderSSR; } });
var fragment_js_1 = require("./fragment.js");
Object.defineProperty(exports, "Fragment", { enumerable: true, get: function () { return fragment_js_1.Fragment; } });
var store_js_1 = require("./store.js");
Object.defineProperty(exports, "Store", { enumerable: true, get: function () { return store_js_1.Store; } });
var context_js_1 = require("./context.js");
Object.defineProperty(exports, "createContext", { enumerable: true, get: function () { return context_js_1.createContext; } });
Object.defineProperty(exports, "useContext", { enumerable: true, get: function () { return context_js_1.useContext; } });
var withStyles_js_1 = require("./withStyles.js");
Object.defineProperty(exports, "withStyles", { enumerable: true, get: function () { return withStyles_js_1.withStyles; } });
var customElementsMode_js_1 = require("./customElementsMode.js");
Object.defineProperty(exports, "defineAsCustomElements", { enumerable: true, get: function () { return customElementsMode_js_1.defineAsCustomElements; } });
// version
var helpers_js_2 = require("./helpers.js");
Object.defineProperty(exports, "printVersion", { enumerable: true, get: function () { return helpers_js_2.printVersion; } });
var version_js_1 = require("./version.js");
Object.defineProperty(exports, "VERSION", { enumerable: true, get: function () { return version_js_1.VERSION; } });
