// nano-jsx/lib/ui/fab.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Fab = void 0;
const component_js_1 = require("../component.js");
const core_js_1 = require("../core.js");
const _config_js_1 = require("./_config.js");
const _helpers_js_1 = require("./_helpers.js");
class Fab extends component_js_1.Component {
    render() {
        const { background = '#6200EE', color = 'white', extended = false, mini = false, center = false, left = false, onClick = () => { } } = this.props;
        const height = mini ? 40 : extended ? 48 : 56;
        const cssHash = (0, core_js_1.strToHash)(extended.toString() + mini.toString() + center.toString() + left.toString());
        const className = `fab-container-${cssHash}`;
        const styles = `
      .${className} {
        ${mini ? 'width: 40px;' : extended ? 'padding: 0px 12px;' : 'width: 56px;'}
        height: ${height}px;
        position: fixed;
        background: ${background};
        border-radius: 36px;
        display: flex;
        align-items: center;
        justify-content: center;
        color: ${color};
        cursor: pointer;

        z-index: ${_config_js_1.zIndex.fab}
        bottom: ${this.props.offsetY ? 16 + this.props.offsetY : 16}px;        
        ${left ? 'left: 16px;' : 'right: 16px;'}
        ${center ? 'transform: translateX(50%); right: 50%;' : ''}
        ${_config_js_1.boxShadow}
        ${_config_js_1.userSelect}
      }
    `;
        (0, _helpers_js_1.addStylesToHead)(styles, cssHash);
        const { children } = this.props;
        return (0, core_js_1.h)('div', { class: className, onClick: (e) => onClick(e) }, children);
    }
}
exports.Fab = Fab;
