// nano-jsx/esm/ui/button.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { h, strToHash } from '../core.js';
import { boxShadow, rippleEffect, userSelect, zIndex } from './_config.js';
import { addStylesToHead, lightenColor } from './_helpers.js';
import { Icon } from './icon.js';
export const Button = (props) => {
    const { children, outlined = false, text = false, background = '#6200ee', color = '#ffffff', style = '', class: className = '', icon } = props, rest = __rest(props, ["children", "outlined", "text", "background", "color", "style", "class", "icon"]);
    const normal = !(outlined || text);
    const bg = normal ? background : '#ffffff';
    const clr = normal ? color : background;
    const hoverClr = normal ? lightenColor(bg, 10) : lightenColor(bg, -10);
    const rippleClr = normal ? lightenColor(bg, 50) : lightenColor(background, 50);
    const cssHash = strToHash(outlined.toString() + text.toString() + bg + clr + style);
    const ripple = rippleEffect(rippleClr, hoverClr);
    const styles = `
    .nano_jsx_button-${cssHash} {
      color: ${clr};
      background: ${bg};
      border-radius: 4px;
      display: inline-flex;
      font-size: 14px;
      padding: 10px 16px;
      margin: 0px 0px 1em 0px;
      text-align: center;
      cursor: pointer;

      ${userSelect}
      

      z-index: ${zIndex.button}

      ${boxShadow}

      border: none;
      text-transform: uppercase;
      box-shadow: 0 0 4px #999;
      outline: none;
    }

    ${ripple.styles}
  `;
    addStylesToHead(styles, cssHash);
    let customStyles = '';
    if (outlined || text) {
        customStyles += 'padding-top: 9px; padding-bottom: 9px; ';
        customStyles += '-webkit-box-shadow: none; -moz-box-shadow: none; box-shadow none; ';
        if (outlined)
            customStyles += `border: 1px ${clr} solid; `;
    }
    customStyles += style;
    return h('button', Object.assign({ class: `nano_jsx_button-${cssHash} ${ripple.class} ${className}`, style: customStyles }, rest), icon ? h(Icon, { style: 'margin-left: -4px; margin-right: 8px; width: 14px; height: 14px;' }, icon) : null, children);
};
